libsolidity/analysis/NameAndTypeResolver.cpp

```diff
@@ -260,20 +260,42 @@ vector<Declaration const*> NameAndTypeResolver::cleanedDeclarations(
 	for (auto it = _declarations.begin(); it != _declarations.end(); ++it)
 	{
 		solAssert(*it, "");
-		// the declaration is functionDefinition while declarations > 1
-		FunctionDefinition const& functionDefinition = dynamic_cast<FunctionDefinition const&>(**it);
-		FunctionType functionType(functionDefinition);
-		for (auto parameter: functionType.parameterTypes() + functionType.returnParameterTypes())
-			if (!parameter)
-				reportFatalDeclarationError(_identifier.location(), "Function type can not be used in this context");
+		// the declaration is functionDefinition or a VariableDeclaration while declarations > 1
+		solAssert(dynamic_cast<FunctionDefinition const*>(*it) || dynamic_cast<VariableDeclaration const*>(*it),
+			"Found overloading involving something not a function or a variable");
+
+		unique_ptr<FunctionType const> functionType {};
+
+		if (FunctionDefinition const* functionDefinition = dynamic_cast<FunctionDefinition const*>(*it))
+		{
+			functionType = unique_ptr<FunctionType const>(new FunctionType(*functionDefinition));
+			for (auto parameter: functionType->parameterTypes() + functionType->returnParameterTypes())
+				if (!parameter)
+					reportFatalDeclarationError(_identifier.location(), "Function type can not be used in this context");
+		}
+		else
+		{
+			VariableDeclaration const* variableDeclaration = dynamic_cast<VariableDeclaration const*>(*it);
+			functionType = unique_ptr<FunctionType const>(new FunctionType(*variableDeclaration));
+		}
+		solAssert(functionType, "failed to determine the function type of the overloaded");
 
 		if (uniqueFunctions.end() == find_if(
 			uniqueFunctions.begin(),
 			uniqueFunctions.end(),
 			[&](Declaration const* d)
 			{
-				FunctionType newFunctionType(dynamic_cast<FunctionDefinition const&>(*d));
-				return functionType.hasEqualArgumentTypes(newFunctionType);
+				if (FunctionDefinition const* functionDefinition = dynamic_cast<FunctionDefinition const*>(d))
```

Comment:
I prefer a function that takes two closures.  One closure takes a `FunctionDefinition const&` as an argument while the other takes `VariableDeclaration const&` as an argument.